Custom font support

Gillibald opened the pull request:
- What does the pull request do?
It introduces support for loading fonts from embedded resources. 
- What is the current behavior?
Only system installed fonts are supported.
- What is the updated/expected behavior with this PR?
You can now define a FontFamily with resource syntax 
- How was the solution implemented (if it's not obvious)?
To speed up loading times and save memory resources are cached. Items of that cache are idetified by a FontFamilyKey that holds a location and optional filename. If you define several FontFamilies with same location and filename it only loads resources once.

To define a FontFamily you can use the commonly used resm scheme. If you point to a resource folder all containing fonts are automaticly discovered. It also introduces filename globbing or so called wildcard support of filenames. That way you can define a FontFamily and automaticly lookup all typefaces in one line.

You can also define fallbacks for your FontFamily. If a font isnt installed on the system, the logic tries to use the fallback values you defined for your FontFamily to render the text.

Some examples of syntax:

Installed font
`
FontFamily="Arial"
`
Installed font with fallbacks
`
FontFamily="Arial, Times New Roman, Comic Sans MS"
`
All font files in a folder
`
FontFamily="resm:DefaultNamespace.Fonts#MyFont"
`
Single font file
`
FontFamily="resm:DefaultNamespace.Fonts.MyFont.ttf#MyFont"
`
All font files in a folder that match expression
`
FontFamily="resm:DefaultNamespace.Fonts.MyFont*.ttf#MyFont"
`

If your font resources are not in your main assembly you have to define the assembly attribute in your uri strings.

To use specific font weights or styles make sure your font is supporting them. If you want to have support for multiple typefaces you have to add them by folder. Your resources would look like this:

```
DefaultNamespace.Fonts.MyFont.MyFont_Regular.ttf
DefaultNamespace.Fonts.MyFont.MyFont_Bold.ttf
DefaultNamespace.Fonts.MyFont.MyFont_Italic.ttf
DefaultNamespace.Fonts.MyFont.MyFont_Italic_Bold.ttf
etc.
```
Then you can use them like so:

```
FontFamily="resm:DefaultNamespace.Fonts.MyFont#MyFont"
FontFamily="resm:DefaultNamespace.Fonts.MyFont#MyFont" FontWeight="Bold"
FontFamily="resm:DefaultNamespace.Fonts.MyFont#MyFont" FontStyle="Italic"
FontFamily="resm:DefaultNamespace.Fonts.MyFont#MyFont" FontWeight="Bold" FontStyle="Italic"

```
I noticed some ui freaze running this under DotNetCore that is gone after everything is loaded. Have to figure out what is causing this.

With this approach we could also support loading fonts from filesystem. Let me know if thats a wanted feature or if we should introduce that later.

Checklist:

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?

See #1313 
See #1564 
See #810

Gillibald committed:
Initial

Gillibald committed:
Direct2D1 experimental

Gillibald committed:
caching

Gillibald committed:
Direct2D1 caching

Gillibald committed:
Skia support

Gillibald committed:
Sample revert

Gillibald committed:
Assets revert

Gillibald committed:
Merge branch 'master' into feature/customFonts

Gillibald committed:
tests fixed

Gillibald committed:
Merge branch 'feature/customFonts' of https://github.com/Gillibald/Avalonia into feature/customFonts

Gillibald committed:
resm filename globbing

Gillibald committed:
res support finished

Gillibald committed:
Finished skia support

Gillibald committed:
Added comments

Gillibald committed:
Removed fonts

Gillibald committed:
Merge branch 'master' into feature/customFonts

Gillibald committed:
Introduce FontFamily.Parse

Gillibald committed:
Rebase for portable.xaml submodule

danwalmsley commented:
@Gillibald great work, i'll try and test this shortly.

Any ideas on how we can support font family lists:

(from msft docs)
fontFamilyNamesList
A string specifying multiple font family names, each separated by a comma (any whitespace following a comma is ignored). The first font family specified serves as the primary font family; subsequent font families serve as fallback families to be used in cases where the primary font family is unavailable or not applicable. For example, "Arial, Century Gothic" specifies Arial as the primary font family, with Century Gothic as the fallback font family.

i.e. 
```
FontFamily="Arial, Verdana, Comic Sans Ms"
```

danwalmsley commented:
I think your PR should change the api of: TextFormatterFactory.CreateFormattedText to accept FontFamily instead of string for name of fontfamily?

After testing your PR I'm having to change that api in lots of places.

Gillibald commented:
In addition to use a fallback for non present fonts we could use a fallback in case a specific typeface isn't present. For example if we set FontWeight to Bold and the main FontFamily has no face to represent that setting we could check if the fallback has one and use that. Not sure how to cache that properly but should be doable.

danwalmsley commented:
I'm getting `System.MissingFieldException: 'Field not found: 'Avalonia.Controls.Primitives.TemplatedControl.FontFamilyProperty'.'` maybe I didn't build something right!

Gillibald commented:
Will have a look at TextFormatterFactory.CreateFormattedText
This PR is tested with DirectD21 and Skia backend so it should just work. Even styles with FontFamily definitions should just work.

danwalmsley commented:
Probably just add an extra overload that takes in fontfamily so we don't break existing api
TextFormatterFactory.CreateFormattedText

Gillibald commented:
Are you sure that TextFormatterFactory.CreateFormattedText is part of this project?

danwalmsley commented:
oops sorry! I was looking at avalonia edit code.

danwalmsley commented:
Note: This PR will not be binary compatible due to change of FontFamily property type.

Gillibald committed:
Minor fixes; Removed res scheme support

Gillibald committed:
Made font API internal
Add FamilyNameList
Add SKTypefaceCollection

Gillibald committed:
FontFamily fallback

Gillibald committed:
Merge branch 'master' into feature/customFonts

Gillibald committed:
First unit tests

Gillibald committed:
Merge branch 'feature/customFonts' of https://github.com/Gillibald/Avalonia into feature/customFonts

Gillibald commented:
Need to add some render tests for DWrite and Skia implementation other than that this is feature complete and can be tested.

Gillibald committed:
Merge branch 'master' into feature/customFonts

Commit:
Merge branch 'master' into feature/customFonts

Commit:
Merge branch 'master' into feature/customFonts

danwalmsley commented:
@Gillibald this still doesn't work for me all fonts get resolved to Consolas or something similar.

![image](https://user-images.githubusercontent.com/4672627/40123786-ec7d839c-591e-11e8-859a-8d889fab4854.png)

Can you contact me on gitter so we can diagnose the issues?

danwalmsley commented on src/Avalonia.Controls/TextBlock.cs:
```diff
@@ -227,7 +227,7 @@ public TextAlignment TextAlignment
         /// <param name="control">The control.</param>
         /// <returns>The font family.</returns>
-        public static string GetFontFamily(Control control)
+        public static FontFamily GetFontFamily(Control control)
```
You must also change the SetFontFamily method. 

```
public static void SetFontFamily(Control control, FontFamily value)
        {
            control.SetValue(FontFamilyProperty, value);
        }
```
currently it throws an exception when this is called.

danwalmsley commented on src/Avalonia.Visuals/Media/FontFamily.cs:
```diff
@@ -0,0 +1,119 @@
+{
+    public class FontFamily
+    {
+        internal static FontFamily Default = new FontFamily("Courier New");
```
Can we use the same font that is currently default? Some apps may be relying on this being consistent.

This should be a font or series of fonts available cross platform, perhaps we can see what WPF default it, and have font fallbacks to similar fonts for Linux, mac and ios / android.

Gillibald replied:
I just used what already was the default just defined it as a static property. This should be used everywhere in our code base when the default is needed.

danwalmsley commented:
I also get `System.Exception: 'resm:AvalonStudio.Assets.SourceSansPro-Regular.ttf;assembly=AvalonStudio#Source Sans Pro is not a valid value for Double.'`

If I try to use a font from a different assembly.

danwalmsley commented:
@Gillibald I wonder is it possible to have it so once a font has been loaded say declared as a resource:
like:
`<FontFamily x:Key="EditorFont">resm:AvalonStudio.Assets.SourceCodePro-Regular.ttf#Source Code Pro</FontFamily>`

It could be then accessed like FontFamily="Source Code Pro" 

where it gets located in the cache?

danwalmsley commented:
Also if I set a FontWeight i.e. Bold it resets the fontfamily to Courier New

Gillibald commented:
I can tests this PR against Avalon Studio if you want. Defining a custom font with full resource string and then reference it just by name at other places is not possible because you can't identify it properly. You still can define your FontFamily in a ResourceDictionary and reference that in your controls.

I can't emaging how you would get an Exception related to a double value.

Gillibald commented:
If you font can't represent the font weight the default is used. That an expected result. You need to load the correct font for your FontWeight and Style

danwalmsley commented:
I'm posting a repro for control catalog, that shows the fonts wont seem to load from separate dll to the main executable.

Repros available here:
https://github.com/AvaloniaUI/Avalonia/tree/repros/custom-fonts-pr

Just load the TextBox page, font isn't resolved, the fonts are embedded in ControlCatalog/Assets/...

It will crash immediately pending the fix to TextBlock.SetFontFamily api. this is an easy change.

Gillibald committed:
Merge branch 'master' into feature/customFonts

Gillibald committed:
TextBlock.SetFontFamily fix

Gillibald commented:
I have tested your repro and could run everything with the fix to TextBlock.SetFontFamily. But i had to fix your resource strings a bit. The Default assembly in your case wasn't ControlCatalog so it could't resolve that. I had to change it like so:

`FontFamily="resm:ControlCatalog.Assets.SourceSansPro-Regular.ttf?assembly=ControlCatalog#Source Sans Pro" 
`

The default assembly is always the entry assembly and thats for example ControlCatalog.Desktop

danwalmsley commented on src/Avalonia.Visuals/Media/Typeface.cs:
```diff
@@ -7,18 +7,8 @@ namespace Avalonia.Media
-            double fontSize,
-            FontStyle style = FontStyle.Normal,
-            FontWeight weight = FontWeight.Normal)
+        public Typeface(FontFamily fontFamily, double fontSize = 12, FontStyle style = FontStyle.Normal,
```
Need to add documentation to this method.

Gillibald replied:
I think I should spend some time on comments in general. There is a lot room for improvements.

danwalmsley commented:
Ok im going to try folder loading now.

danwalmsley commented:
This is looking really great btw, hopefully we can get it merged soon.

Gillibald commented:
I have updated the description of this PR to explain how to use multiple typfaces. Have to write some docs.

danwalmsley commented:
Ok pretty simple just use *. i.e.
`<FontFamily x:Key="InputFont">resm:ControlCatalog.Assets.SourceSansPro*.ttf?assembly=ControlCatalog#Source Sans Pro</FontFamily>`

danwalmsley commented:
ok this is all working great, i'll see how things run on Linux in the morning.

Commit:
Merge branch 'master' into feature/customFonts

kekekeks commented on src/Avalonia.Visuals/Media/Fonts/FontFamilyLoader.cs:
```diff
@@ -0,0 +1,191 @@
+        /// <returns></returns>
+        private static IEnumerable<FontResource> GetFontResourcesByLocation(Uri location)
+        {
+            var assembly = GetAssembly(location);
```
You shouldn't be accessing reflection API directly. That's what IAssetLoader is there for.

kekekeks commented on tests/Avalonia.RenderTests/Controls/BorderTests.cs:
```diff
@@ -204,7 +204,7 @@ public BorderTests()
                     {
                         Text = "Foo",
                         Background = Brushes.Red,
-                        FontFamily = "Segoe UI",
```
This is a breaking change. Please,  add an implicit convertion operator to the FontFamily class

Gillibald replied:
Okay that's a good idea

Gillibald replied:
Have implemented implicit conversion

danwalmsley replied:
If there is no an implicit conversion (I couldn't where is it btw?) then this change is no longer necessary, please revert all the FontFamily = new FontFamily( "") changes back to FontFamily = "".

Gillibald replied:
I am not sure if I should revert everything to use string. Implicit conversion is just needed for backwards compatibility.

danwalmsley replied:
My though process was that then this PR only shows the necessary changes, however i'll let you make the best judgement on this one :)

danwalmsley commented on src/Shared/PlatformSupport/AssetLoader.cs:
```diff
@@ -218,5 +216,5 @@ public AssemblyDescriptor(Assembly assembly)
             public string Name { get; }
         }
-    }
+    }  
```
Remove whitespace added here.

danwalmsley commented on src/Windows/Avalonia.Direct2D1/Direct2D1Platform.cs:
```diff
@@ -117,7 +117,7 @@ public IBitmapImpl CreateBitmap(int width, int height)
 
         public IFormattedTextImpl CreateFormattedText(
-            string text,
+            string text,           
```
remove whitespace change.

Gillibald committed:
Requested changes

Gillibald committed:
Merge branch 'feature/customFonts' of https://github.com/Gillibald/Avalonia into feature/customFonts

Gillibald committed:
Merge branch 'master' into feature/customFonts

Gillibald committed:
IAssetLoader changes
Rename resource to asset

Gillibald committed:
Fixed unit tests

Gillibald commented:
I have added a GetAssets method to IAssetLoader and renamed resource to asset to keep the naming consistent. The changes to IAssetLoader should probably have its own PR. Will do that shortly and will revert the changes I made until that got merged.

danwalmsley commented:
I will have a look soon. 😀

danwalmsley commented on src/Shared/PlatformSupport/AssetLoader.cs:
```diff
@@ -83,7 +83,7 @@ public bool Exists(Uri uri, Uri baseUri = null)
         /// The resource was not found.
         /// </exception>
-        public Tuple<Stream, Assembly> OpenAndGetAssembly(Uri uri, Uri baseUri = null)
+        public (Stream Stream, Assembly Assembly) OpenAndGetAssembly(Uri uri, Uri baseUri = null)
```
Nit:
Could we use `camelCase` for tuple values. It would look odd when a tuple is deconstructed and you end up with locals that start with capitals.

Gillibald replied:
I will change that.

danwalmsley commented on src/Shared/PlatformSupport/AssetLoader.cs:
```diff
@@ -92,9 +92,18 @@ public bool Exists(Uri uri, Uri baseUri = null)
+                   Enumerable.Empty<(string AbsolutePath, Assembly Assembly)>();
         }
 
+
```
unnecessary whitespace.

danwalmsley commented on src/Windows/Avalonia.Direct2D1/Media/DWriteResourceFontFileStream.cs:
```diff
@@ -0,0 +1,85 @@
+                _stream.Position = fileOffset;
+
+                fragmentStart = _stream.PositionPointer;
+
```
Nit: unnecessary line space

danwalmsley reviewed (changes requested):
Some minor nits

danwalmsley commented on src/Avalonia.Visuals/Media/FontFamily.cs:
```diff
@@ -79,9 +79,23 @@ internal FamilyNameCollection FamilyNames
+        /// Implicit conversion of FontFamily to string
+        /// </summary>
+        /// <param name="fontFamily"></param>
+        public static implicit operator string(FontFamily fontFamily)
```
I think this should have been the other way around, i.e. string to FontFamily.

Gillibald replied:
The FontFamily was defined as a string so if someone still has some code with FontFamily = "MyFont" it still works. The other way isn't really needed. There is already ToString

grokys commented on src/Avalonia.Visuals/Properties/AssemblyInfo.cs:
```diff
@@ -10,5 +10,7 @@
 [assembly: XmlnsDefinition("https://github.com/avaloniaui", "Avalonia.Media")]
 [assembly: XmlnsDefinition("https://github.com/avaloniaui", "Avalonia")]
 
+[assembly: InternalsVisibleTo("Avalonia.Direct2D1")]
```
We shouldn't really be doing this - it should be possible for people to write their own rendering backends without us adding an `InternalsVisibleTo` attribute.

grokys commented on src/Avalonia.Visuals/Media/Fonts/FontAsset.cs:
```diff
@@ -0,0 +1,27 @@
+        /// Initializes a new instance of the <see cref="FontAsset"/> class.
+        /// </summary>
+        /// <param name="source">The source.</param>
+        public FontAsset(Uri source)
```
This class just contains a single `Uri` - why is this abstraction needed? Why not just pass a `Uri`?

grokys reviewed (changes requested):
Looking really good! Just one question and one change I think we need.

Gillibald committed:
Requested changes

Gillibald committed:
Spacing fix

Gillibald commented:
Dont know if I am happy with IFontFamily. Pls let me know your thoughts.

Gillibald committed:
New asset loader
Added unit test for wildcard

Gillibald committed:
Merge branch 'master' into feature/customFonts

Gillibald committed:
Cleanup

Gillibald committed:
Unit test fix

danwalmsley reviewed (approved):
LGTM 👍

grokys commented on src/Avalonia.Visuals/Media/FontFamily.cs:
```diff
@@ -0,0 +1,156 @@
+
+        public string Name => FamilyNames.PrimaryFamilyName;
+      
+        FamilyNameCollection IFontFamily.FamilyNames => FamilyNames;
```
I don't understand this: `IFontFamily` exposes `Name`, `FamilyNames` and `Key` which are implemented here as explicit interface methods and forwarded to `internal` properties? Why not just make these properties public? That would have the same effect, no?

And why is `IFontFamily` needed in that case? It doesn't look like there can be any other implementation of it, as the `TextBlock.FontFamily` property is of type `FontFamily`. What use-case would having another implementation of `FontFamily` solve?

Gillibald replied:
Have removed the interface. It was ment to hide FamilyNames and Key for 3rd Party. The internal properties just made it easier to access them without casting.

grokys reviewed (changes requested):
Hey, I don't understand the `IFontFamily` changes - I know we talked about this in gitter and I thought that it was supposed to hide some implementation details, but I can't see how it's doing that here.

Gillibald committed:
Removed IFontFamily
Some code style fixes

Gillibald commented on samples/ControlCatalog/ControlCatalog.csproj:
```diff
@@ -11,6 +11,18 @@
     <EmbeddedResource Include="Assets\*" />
   </ItemGroup>
+  <ItemGroup>
+    <None Remove="Assets\Fonts\SourceSansPro-Bold.ttf" />
```
I have no idea why visual studio added them. They are not needed.

Gillibald committed:
ControlCatalog .csproj fix

Commit:
Merge branch 'master' into feature/customFonts

jpgarza93 commented:
Can we add this to the docs? I am having trouble finding info on how to add a custom font.

timunie commented:
@jpgarza93 if you now understood what to do, you are welcome to provide a PR to the docs: https://github.com/AvaloniaUI/Documentation

jpgarza93 commented:
@timunie I didn't understand :( but once I do I will, thanks
